### ISSUE TITLE
Custom opportunities matrix

### ISSUE THREAD

**Issue by gevro** [AlexandrovLab/SigProfilerExtractor#225]
Hi, Is there a way to specify a custom opportunities matrix instead of a genome reference version?

Thank you.

**Comment by marcos-diazg** [comment]
Hi @gevro,

Thanks for your question! At the moment, only COSMIC reference signatures can be used as opportunity matrices in SigProfilerExtractor. However, if you are interested in decomposing _de novo_ extracted signatures or assigning signatures to samples using custom reference signatures, you can use [SigProfilerAssignment](https://github.com/AlexandrovLab/SigProfilerAssignment/) and provide your custom matrix to the `signature_database` parameter.

I hope this helps. Best,

Marcos

**Comment by gevro** [comment]
Hi Marcos, Bu opportunities, I actually meant the term in the way that sigfit uses it. In other words, if sampling from a biased, trinucleotide background, then an opportunities matrix normalizes for that. I'm not referring to a signatures reference like cosmic.

Is that option possible?
Thanks!

**Comment by marcos-diazg** [comment]
Hi again @gevro,

Thanks for clarifying. We provide precalculated renormalized matrices for different reference genomes, as well as exome and genome, by setting the corresponding parameters, both in SigProfilerExtractor and SigProfilerAssignment. This is how our tools currently handle the differences in the trinucleotide backgrounds. I hope this helps!

Best,

Marcos

**Comment by gevro** [comment]
Thanks. Is there any way to provide a custom trinuckeotide background that is different from one of the pre-computed reference genomes?

**Comment by marcos-diazg** [comment]
No, unfortunately, this is not currently possible. But you can use your custom trinucleotide background to generate your custom set of signatures/mutations and provide them to our SigProfilerAssignment tool.

**Comment by gevro** [comment]
Regarding this, I realize I'm not completely clear how to do this.

My goal is to extract de novo signatures from my samples. These samples are from a biased fraction of the GRCh38 genome (similar to an exome).

After making a matrix with SigProfilerMatrixGenerator from my vcf files, I run SigProfilerExtractor.

However, SigProfilerExtractor requires an 'opportunity_genome' parameter to be defined. It also says the 'reference_genome' parameter is available, but only for vcf input.

Questions:
1. If I use a matrix input created by SigProfilerMatrixGenerator, does SigProfiler know how to find the 'reference_genome' from which the VCF files were from (i.e. 'GRCh38)?
2. What does 'opportunity_genome' do exactly? Does it correct the mutational spectrum from 'reference_genome' into 'opportunity_genome'? Even if I set this to 'GRCh38', the result would not actually correspond to 'GRCh38' due to the biased genomic regions of my data, even though SigProfiler thinks the output is GRCh38.
3. Is the solution to run SigProfilerExtractor with 'opportunity_genome' = GRCh38, and then correct the resulting signatures from my biased regions of the genome to GRCh38? If so, is there a tool to transform signatures given a ratio of source genome to destination genome that is fully customizable, i.e. not just GRCh37, 38, etc

Basically, this is a similar situation to a custom targeted capture panel, like an exome.
Thanks so much for any advice!

**Comment by marcos-diazg** [comment]
Hi @gevro,

Sorry for the delay in getting back to you. Answering your questions:

1. The parameter `reference_genome` is, indeed, passed to SigProfilerMatrixGenerator internally to generate the input mutational matrix. If you already have your mutational matrix to use as input for SigProfilerExtractor, this parameter is not used.

2. The parameter `opportunity_genome` controls the version of the COSMIC signatures that is passed internally to SigProfilerAssignment for the decomposition and assignment, after the _de novo_ extraction of mutational signatures. Our strategy to deal with different genomic regions and their different trinucleotide frequencies is to precalculate different versions of the COSMIC signatures, instead of modifying/correcting the input mutational matrix.

3. In your particular case, you will need to calculate specific versions of the COSMIC signatures for your particular region of the genome. Unfortunately, we don't have a tool to transform signatures. However, you can use SigProfilerMatrixGenerator to calculate the trinucleotide frequencies in your genomic region of interest (you will need a BED file for this), using the following [script](https://github.com/AlexandrovLab/SigProfilerMatrixGenerator/blob/master/SigProfilerMatrixGenerator/scripts/save_context_distribution.py). After that, you can use the trinucleotide frequencies in the whole genome (available [here](https://github.com/AlexandrovLab/SigProfilerMatrixGenerator/tree/master/SigProfilerMatrixGenerator/references/chromosomes/context_distributions)) to calculate a ratio and modified the default COSMIC signatures.

I hope this helps, and happy to follow up by email (mdiazgay@ucsd.edu) if you have further questions.

**Comment by gevro** [comment]
Thankyou!!
